Migration Guide for Flutter web slotted rendering.

ditman opened the pull request:
This change adds a Migration Guide for the new Flutter web slotted rendering, explaining how the code is changing and what users need to do for their web apps to keep working as normal.

Fixes: https://github.com/flutter/flutter/issues/80524

yjbanov commented on src/docs/release/breaking-changes/platform-views-using-html-slots-web.md:
```diff
@@ -0,0 +1,196 @@
+---
+title: Using HTML slots to render platform views in the web
+description: iframes in Flutter web used to reload, because of the way some DOM operations were made. A change in the way Flutter web apps render Platform Views makes them much more stable (preventing iframe reloads, and other problems with video tags or forms potentially losing their state).
```
Remove "much more". It wasn't stable and now it simply is :)

ditman replied:
Hah, yep. Removed!

yjbanov commented on src/docs/release/breaking-changes/platform-views-using-html-slots-web.md:
```diff
@@ -0,0 +1,196 @@
+Flutter now renders Platform Views using [slot elements][] inside of a single,
+app-wide [Shadow Root][]. Slot elements can be added/removed/moved around the
+Shadow DOM without affecting the underlying slotted content (which is rendered
+in a constant location)
```
This discusses an implementation detail, which may be useful in the "Context" section below. In the summary I think the developers would be most interested in what happens to the HTML DOM that _they_ created, so I'd suggest that we talk about how we put their platform views into the root element, instead of injecting it into the Flutter's internal DOM structures.

ditman replied:
I moved this to the `context` section.

yjbanov commented on src/docs/release/breaking-changes/platform-views-using-html-slots-web.md:
```diff
@@ -0,0 +1,196 @@
+
+## Context
+
+The Flutter framework frequently tweaks its Render Tree to optimize the paint
```
I'm not sure if "render tree" is worth capitalizing.

ditman replied:
Capitalization Of Things, The Bane Of My Existence :P

ditman replied:
(lower-cased :P)

yjbanov commented on src/docs/release/breaking-changes/platform-views-using-html-slots-web.md:
```diff
@@ -0,0 +1,196 @@
+
+* make Platform Views behave in a stable way in Flutter web.
+* unify the way Platform Views were rendered in the web for both rendering
+backends (`html` and `canvaskit`).
```
And:

* provide a predictable location in the DOM to allow developer to reliably use CSS to style HTML inside platform views, and use other standard DOM API, such as `querySelector`, and `getElementById`.

yjbanov commented on src/docs/release/breaking-changes/platform-views-using-html-slots-web.md:
```diff
@@ -0,0 +1,196 @@
+operations that are ultimately made per frame. In the web, these Render Tree
+changes often result in DOM operations.
+
+Flutter web used to render its Platform Views (`HtmlElementView` widgets)
```
I'd probably also not capitalize "platform views"

yjbanov commented on src/docs/release/breaking-changes/platform-views-using-html-slots-web.md:
```diff
@@ -0,0 +1,196 @@
+
+## Description of change
+
+A Flutter web app is now rendered inside a common Shadow Root in which slot
```
I'd not capitalize "shadow root", but I would link to MDN or some other reliable source that explains what it is.

yjbanov commented on src/docs/release/breaking-changes/platform-views-using-html-slots-web.md:
```diff
@@ -0,0 +1,196 @@
+A Flutter web app is now rendered inside a common Shadow Root in which slot
+elements are used to represent platform views. The actual content of each
+platform view is rendered *outside* of the Shadow Root, as a child of the Shadow
+Root host.
```
I think we can gzip this sentence a bit: "...is rendered as a sibling of the shadow root"

yjbanov replied:
Or, since we're keeping `<flt-glass-pane>` we could just say "direct child of `<flt-glass-pane>`"

yjbanov commented on src/docs/release/breaking-changes/platform-views-using-html-slots-web.md:
```diff
@@ -0,0 +1,196 @@
+
+### Code
+
+The content returned by the users' Platform View Factory used to be resized and
```
nit: user's

yjbanov replied:
nit: s/resized/sized/

yjbanov commented on src/docs/release/breaking-changes/platform-views-using-html-slots-web.md:
```diff
@@ -0,0 +1,196 @@
+### Code
+
+The content returned by the users' Platform View Factory used to be resized and
+positioned by the framework, but not anymore. Ensure that your returned
```
I'd insert another sentence between "but not anymore." and "Ensure...":

...but not anymore. Instead, Flutter sizes and positions `<flt-platform-view>`, which is the parent of the content. Ensure...

yjbanov replied:
I'd also replace "Ensure" with "To ensure". Maybe the developer doesn't want to fill the whole space. They only need to set size to 100% if they want to.

yjbanov commented on src/docs/release/breaking-changes/platform-views-using-html-slots-web.md:
````diff
@@ -0,0 +1,196 @@
+<!-- skip -->
+```bash
+Height of Platform View type: [$viewType] may not be set. Defaulting to `height: 100%`.
+Set `style.height` to any appropriate value to stop this message.
````
Hmm, do we need this? Why can't I use relative layout (flexbox, etc) to fill the space?

ditman replied:
If you use any other way of filling the space, you need to set height/width to `auto`, or anything that is not empty. In the engine the only thing we look at is that width/height is set.

ditman replied:
I'll add a "But I don't want to set width: 100% in my platform view" section

yjbanov commented on src/docs/release/breaking-changes/platform-views-using-html-slots-web.md:
````diff
@@ -0,0 +1,196 @@
+<!-- skip -->
+```bash
+Width of Platform View type: [$viewType] may not be set. Defaulting to `width: 100%`.
+Set `style.width` to any appropriate value to stop this message.
````
same here

yjbanov commented on src/docs/release/breaking-changes/platform-views-using-html-slots-web.md:
```diff
@@ -0,0 +1,196 @@
+
+### Tests
+
+Unfortunately, there are no selectors that let one _pierce_ through the Shadow
```
s/Unfortunately/By design/ (this is why we use it 😈)

yjbanov commented on src/docs/release/breaking-changes/platform-views-using-html-slots-web.md:
```diff
@@ -0,0 +1,196 @@
+
+### Tests
+
+Unfortunately, there are no selectors that let one _pierce_ through the Shadow
+DOM boundary to locate elements underneath it, so code that once peeked inside
+of the `flt-glass-pane`, needs to be made Shadow DOM aware. Using the html
```
nit: remove comma before "needs"

sfshaza2 replied:
```suggestion
By design, there are no selectors that let one _pierce_ through the shadow
DOM boundary to locate elements underneath it, so code that once peeked inside
of the `flt-glass-pane` needs to be made shadow DOM aware. Using the html
```

yjbanov commented on src/docs/release/breaking-changes/platform-views-using-html-slots-web.md:
````diff
@@ -0,0 +1,196 @@
+<!-- skip -->
+```dart
+final shadowRoot = document.querySelector('flt-glass-pane')?.shadowRoot!;
+final element = shadowRoot.querySelector('#some-element');
````
Why is this necessary? After the migration `#some-element` will no longer be inside the shadow DOM, only the slot. We don't want people looking into Flutter's private DOM structure. It's unpredictable and unstable.

ditman replied:
This is what most of the tests that failed were doing. We can document the fact that now they shouldn't drill inside the shadowDOM, as mentioned above, that's probably better!

ditman replied:
Removed this, warned people against looking into the shadowDom of the app.

sfshaza2 commented on src/docs/release/breaking-changes/platform-views-using-html-slots-web.md:
```diff
@@ -0,0 +1,196 @@
+
+## Description of change
+
+A Flutter web app is now rendered inside a common Shadow Root in which slot
+elements are used to represent platform views. The actual content of each
+platform view is rendered *outside* of the Shadow Root, as a child of the Shadow
+Root host.
```
I'm suggesting a change that corresponds to Yegor's comment However, he might have intended for that second sentence to be deleted? @yjbanov?

```suggestion
A Flutter web app is now rendered inside a common [shadow root][] in which slot
elements represent platform views. The actual content of each
platform view is rendered *outside* of the shadow root, as a child of the shadow
root host.

[shadow root]: https://www.html5rocks.com/en/tutorials/webcomponents/shadowdom-201/
```

sfshaza2 commented on src/docs/release/breaking-changes/platform-views-using-html-slots-web.md:
````diff
@@ -0,0 +1,196 @@
+...
+```
+
+Now, when the framework needs to move DOM nodes around, it'll operate over
+`flt-platform-view-slot`s, which only contain the `slot` tag that "projects" the
+content defined in the `flt-platform-view` tags outside of the Shadow Root
+(which never move), thus preventing the reloads.
````
```suggestion
After this change, when the framework needs to move DOM nodes around,
it operates over`flt-platform-view-slot`s, which only contain the `slot` tag that "projects" the
content defined in the `flt-platform-view` tags outside of the shadow root
(which never move), thus preventing the reloads.
```

sfshaza2 commented on src/docs/release/breaking-changes/platform-views-using-html-slots-web.md:
```diff
@@ -0,0 +1,196 @@
+content defined in the `flt-platform-view` tags outside of the Shadow Root
+(which never move), thus preventing the reloads.
+
+From an app's perspective, this change is transparent, **however** this was
+deemed a _breaking change_ because tests that make assumptions about the DOM of
+a Flutter web app can break.
```
```suggestion
From an app's perspective, this change is transparent. **However**, this is
considered a _breaking change_ because some tests make assumptions
about how the DOM of a Flutter web app might break.
```

sfshaza2 commented on src/docs/release/breaking-changes/platform-views-using-html-slots-web.md:
```diff
@@ -0,0 +1,196 @@
+
+### Code
+
+The content returned by the users' Platform View Factory used to be resized and
+positioned by the framework, but not anymore. Ensure that your returned
+`html.Element` takes the whole space allocated for it by the framework, by
+setting its `style.width` and `style.height` properties to `'100%'`:
```
```suggestion
Previously, the content returned by [`PlatformViewFactory`][] was resized and
positioned by the framework. Instead, Flutter now sizes and positions
`<flt-platform-view>`, which is the parent of the content.
To ensure the `html.Element` that you return takes the whole space
allocated to it, set its `style.width` and `style.height` properties to `'100%'`:

[`PlatformViewFactory`]: {{site.api}}/javadoc/index.html?io/flutter/plugin/platform/PlatformView.html
```

sfshaza2 commented on src/docs/release/breaking-changes/platform-views-using-html-slots-web.md:
````diff
@@ -0,0 +1,196 @@
+});
+```
+
+Otherwise the engine will print a warning message to the console similar to:
````
```suggestion
Otherwise, the engine prints a warning message to the console similar to:
```

sfshaza2 commented on src/docs/release/breaking-changes/platform-views-using-html-slots-web.md:
```diff
@@ -0,0 +1,196 @@
+
+## References
+
+Design Doc:
```
```suggestion
Design doc:
```

sfshaza2 reviewed (changes requested):
Yegor made some comments that I turned into official GitHub suggestions. He also had some questions as to the design, and I will leave that to the two of you to resolve.  :)

ditman commented:
Thanks @sfshaza2 for the suggestions! I was halfway through applying by hand Yegor's suggestions so I took the ones from you that wouldn't directly conflict with my changes :P I've manually applied your changes to the first half of the doc, I think I got'em'all!

sfshaza2 reviewed (approved):
LGTM, but I will let Yegor have the final say.

ditman commented:
I've added a few more lines here and there, I think I addressed all of the comments. @yjbanov I've re-tagged you to PTAL. Once we're happy with the content, I'll ping @sfshaza2 for a last review!

yjbanov reviewed (approved):
![lgtm](https://user-images.githubusercontent.com/211513/32384589-a60f0e74-c078-11e7-9bc1-e5b5287aea9d.png)
